src/probnum/diffeq/callbacks/_discrete_callback.py

```diff
@@ -0,0 +1,33 @@
+"""Discrete interventions."""
+
+from typing import Callable, Union
+
+import probnum  # pylint: disable="unused-import"
+from probnum.diffeq.callbacks import _callback
+
+
+class DiscreteCallback(_callback.ODESolverCallback):
+    """Handle discrete events in an ODE solver.
+
+    A discrete event can be a time-stamp that must be included in the locations. It can
+    also be any event for which it is possible to write down a condition that evaluates
+    to `True` or `False`. If a condition evaluates to `True`, the current state can be
+    modified.
+    """
+
+    # New init because condition() types are more specific.
+    def __init__(
+        self,
+        replace: Callable[
```

Comment:
I went for `replace` (previously: `modify`), because (i) this is less suggestive of in-place operations, and (ii) the most convenient way of implementing this function may be with `dataclasses.replace(state, rv=something)`.